### PR TITLE
Filter build orders by responsible owner (#4383)

### LINKED ISSUE #4383: [FR] Filter Build Orders by Assigned User/Group
### Please verify that this feature request has NOT been suggested before.

- [X] I checked and didn't find a similar feature request

### Problem statement

I am using Groups to assign build orders to different manufacturing departments (like "CNC", "Composites", "3D Printing", "Electronics" etc). Currently I don't see a filter option to only show build orders that are assigned to a specific user or group other than yourself.

### Suggested solution

It would be very useful to have an additional filter in the build orders list to filter by `assigned to:` where users and groups can be selected.
Or maybe split into two filters: `assigned to user:` and `assigned to group:` - otherwise the options drop-down would be confusing and possibly crowded.

I think this would live here, right?
https://github.com/inventree/InvenTree/blob/75ca325892a09b6836250b863399649234d95d66/InvenTree/templates/js/translated/table_filters.js#L339-L359

is there a global variable already to access a list of users and groups? In case not, where should that be created best? Right there in `table_filters.js` from an api call, or maybe somewhere more central?

### Describe alternatives you've considered

none that I can think of really - other than sorting the table by `Assigned to` column. But that is not so helpful with huge projects and many active build orders.

### Examples of other systems

_No response_

### Do you want to develop this?

- [X] I want to develop this.

### PR DESCRIPTION
**Pull request by simonkuehling**
I have implemented the API+UI filters for filtering build orders by their responsible users/groups as discussed in #4383.

Filtering by a user also shows build orders assigned to groups he is part of - like the `assigned_to_me` filter does.
Along the way I added a user/group icon to the `Responsible` column, which is taken from the sales orders table.

Full table including the icons:

Bildschirmfoto vom 2023-02-24 12-40-39

Filtered by a responsible group:

Bildschirmfoto vom 2023-02-24 12-41-04

Filtered by a responsible user:

Bildschirmfoto vom 2023-02-24 12-41-30

Closes #4383

### CODE REVIEW

File: InvenTree/build/api.py [hunk 1]
```diff
@@ -65,6 +66,21 @@ def filter_assigned_to_me(self, queryset, name, value):
 
         return queryset
 
+    assigned_to = rest_filters.NumberFilter(label='responsible', method='filter_responsible')
+
+    def filter_responsible(self, queryset, name, value):
+        """Filter by orders which are assigned to the specified owner."""
+        owners = []
+        owners.append(Owner.objects.get(pk=value))
```
Comment (SchrodingersGat) [inline InvenTree/build/api.py:74]: You should use `Owner.objects.filter()` - `get()` can raise an exception if the matching PK does not exist

```python
owners = list(Owner.objects.filter(pk=value))
if len(owners) > 0 and owners[0].label() == 'user':
    ...
```

File: InvenTree/templates/js/translated/build.js [hunk 2]
```diff
@@ -2694,11 +2694,19 @@ function loadBuildTable(table, options) {
...
-                    if (value) {
-                        return row.responsible_detail.name;
-                    } else {
+                    if (!row.responsible_detail) {
                         return '-';
                     }
+
+                    var html = row.responsible_detail.name;
+
+                    if (row.responsible_detail.label == 'group') {
```
Comment (SchrodingersGat) [inline InvenTree/templates/js/translated/build.js:2703]: Nice addition :)

File: InvenTree/templates/js/translated/table_filters.js [hunk 3]
```diff
@@ -355,6 +368,10 @@ function getAvailableTableFilters(tableKey) {
                 type: 'bool',
                 title: '{% trans "Assigned to me" %}',
             },
+            assigned_to: {
+                title: '{% trans "Responsible" %}',
+                options: ownersList,
```
Comment (SchrodingersGat) [inline InvenTree/templates/js/translated/table_filters.js:373]: This is a nice approach, but the API request is going to be made every time the table is rendered (even when not necessary).

I'd say that we want to be able to dynamically fill out the "options" attribute for a table filter when the filter is added. This way that API request only fires when the user selects "owner" from the drop-down menu of available filters. Will need to extend the functionality of `table_filters.js` to achieve this.

**Comment by SchrodingersGat** [comment]
Also please ensure that `InvenTree/api_version.py` gets updated accordingly :)

File: InvenTree/build/api.py (hunk 1 above)
Comment (simonkuehling) [reply InvenTree/build/api.py:74]: thanks for the tip!

But using `User.objects.get()` should be fine, as it is already verified to exist by the first owners query, right?
Or is it generally favorable to use `.filter()` here as well?

Comment (SchrodingersGat) [reply InvenTree/build/api.py:74]: >  as it is already verified to exist by the first owners query, right?

I'm not 100% sure about that - you should check by trying to filter with a bad ID. If it does do that check for you, it'd be fine to use get

File: InvenTree/templates/js/translated/build.js (hunk 2 above)
Comment (simonkuehling) [reply InvenTree/templates/js/translated/build.js:2703]: thanks!

File: InvenTree/templates/js/translated/table_filters.js (hunk 3 above)
Comment (simonkuehling) [reply InvenTree/templates/js/translated/table_filters.js:373]: well, I thought about that, too... We could change the value for this selection option into a function callback in `table_filter.js --> getAvailableTableFilters()`, like so:

```js
[...]
           assigned_to: {
              title: '{% trans "Responsible" %}',
              options: function () { 
                  var ownersList = {};
                  inventreeGet(`/api/user/owner/`, {}, {
                      async: false,
                      success: function(response) {
                          for (key in response) {
                              var owner = response[key];
                              ownersList[owner.pk] = {
                                  key: owner.pk,
                                  value: `${owner.name} (${owner.label})`,
                              };
                          }
                      }
                  });
                  return ownersList;
              },
          },
[...]
```
and then extend the html generation in `filter.js --> setupFilterList()` to run that callback like this:
```js
        if (options instanceof Function){
            // run callback function for filter options
            options = options();
        }
```
Now we can use either a `list` as before, or a `function()` for the filter options.
I tried that and it works flawlessly - should I commit this?

File: InvenTree/build/api.py (hunk 1 above)
Comment (simonkuehling) [reply InvenTree/build/api.py:74]: looking good: any bad ID is successfully discarded by the first `Owner.objects.filter(pk=value)` already, so I'm happy with a `get()` here...

File: InvenTree/templates/js/translated/table_filters.js (hunk 3 above)
Comment (simonkuehling) [reply InvenTree/templates/js/translated/table_filters.js]: I added the proposed solution as a commit for better overview as https://github.com/inventree/InvenTree/pull/4408/commits/acc31a20e1fed4dba11726a91760ab505047fb3e - let me know if you like it or not, I'm open for alternatives as well...

**Comment by SchrodingersGat** [comment]
@simonkuehling nice, seems to work as desired. Thanks for this contribution :)

**Pull request merged**